lib/src/sync-process/sync-message-port.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {strict as assert} from 'assert';
+import {EventEmitter} from 'events';
+import {
+  receiveMessageOnPort,
+  MessageChannel,
+  MessagePort,
+  TransferListItem,
+} from 'worker_threads';
+
+// TODO(nweiz): Make this its own package.
+
+/**
+ * An enum of possible states for the shared buffer that two `SyncMessagePort`s
+ * use to communicate.
+ */
+enum BufferState {
+  /**
+   * The initial state. When an endpoint is ready to receive messages, it'll set
+   * the buffer to this state so that it can use `Atomics.wait()` to be notified
+   * when it switches to `MessageSent`.
+   */
+  AwaitingMessage,
+  /**
+   * The state indicating that a message has been sent. Whenever an endpoint
+   * sends a message, it'll set the buffer to this state so that the other
+   * endpoint's `Atomics.wait()` call terminates.
+   */
+  MessageSent,
+  /**
+   * The state indicating that the channel has been closed. This never
+   * transitions to any other states.
+   */
+  Closed,
+}
+
+/**
+ * A communication port that can receive messages synchronously from another
+ * `SyncMessagePort`.
+ *
+ * This also emits the same asynchronous events as `MessagePort`.
+ */
+export class SyncMessagePort extends EventEmitter {
+  /** Creates a channel whose ports can be passed to `new SyncMessagePort()`. */
+  static createChannel(): MessageChannel {
+    const channel = new MessageChannel();
+    // Four bytes is the minimum necessary to use `Atomics.wait()`.
+    const buffer = new SharedArrayBuffer(4);
+
+    // Queue up messages on each port so the caller doesn't have to explicitly
+    // pass the buffer around along with them.
+    channel.port1.postMessage(buffer);
+    channel.port2.postMessage(buffer);
+    return channel;
+  }
+
+  /**
+   * An Int32 view of the shared buffer.
+   *
+   * Each port sets this to 0 before checking for new messages in
```

Comment:
Consider using the actual BufferState names instead of the underlying numbers in this comment.